File: javascript/src/providers/Dropmail.js
import { TempMailProvider } from '../base.js';
import { Message, MessageDetail } from '../models.js';
import { TempMailError, NotFoundError } from '../errors.js';

const TOKEN_URL = 'https://dropmail.me/api/token/generate';
const PADDLE_OCR_URL = 'https://mamamacjdjj-padle-predict.hf.space/predict';

/**
 * Built-in PaddleOCR solver via HuggingFace space.
 * Tries up to 3 times, returns trimmed text on success or null.
 * @param {ArrayBuffer} imgBytes - captcha image bytes
 * @returns {Promise<string|null>}
 */
export async function paddleOcrSolver(imgBytes) {
  for (let attempt = 0; attempt < 3; attempt++) {
    try {
      const form = new FormData();
      form.append('file', new Blob([imgBytes], { type: 'image/png' }), 'cap.png');
      const ocrRes = await fetch(PADDLE_OCR_URL, { method: 'POST', body: form });
      const ocrData = await ocrRes.json();
      const results = ocrData.results || [];
      if (results.length > 0 && results[0].confidence >= 0.7) {
        return results[0].text.trim();
      }
    } catch { /* ignore */ }
  }
  return null;
}

/**
 * Dropmail.me provider — GraphQL API.
 */
export class Dropmail extends TempMailProvider {
  /** @param {{ randomUA?: boolean, useCookies?: boolean, solvers?: Array<function(ArrayBuffer): Promise<string|null>> }} [opts] */
  constructor(opts = {}) {
    super('dropmail', { randomUA: opts.randomUA, useCookies: opts.useCookies });
    /** @type {Array<function(ArrayBuffer): Promise<string|null>>|null} */
    this._solvers = opts.solvers || null;
    /** @type {string|null} */
    this._token = null;
    /** @type {string|null} */
    this._endpoint = null;
    /** @type {string|null} */
    this._sessionId = null;
    /** @type {string|null} */
    this._addressId = null;
    /** @type {string|null} */
    this._email = null;
  }

  /**
   * Execute a GraphQL query/mutation.
   * @param {string} query
   * @param {object} [variables]
   * @returns {Promise<object>}
   */
  async _graphql(query, variables = {}) {
    if (!this._endpoint) throw new TempMailError('Not initialized. Call generateEmail first.', this.name);

    const res = await this._fetch(this._endpoint, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ query, variables }),
    });

    if (!res.ok) throw new TempMailError(`HTTP ${res.status}`, this.name);
    const data = await res.json();

    if (data.errors && data.errors.length > 0) {
      throw new TempMailError(data.errors[0].message, this.name);
    }

    return data.data;
  }

  /**
   * Solve captcha and get 90d token.
   * Uses this.http.request() (same cookie jar) for Dropmail steps; plain fetch for PaddleOCR.
   * @param {object} captcha - captcha object from 402 response
   * @returns {Promise<string|null>} token or null on failure
   */
  async _solveCaptchaAndGetToken(captcha) {
    const { v = '3', nonce = '', key = '', _sig: sig = '' } = captcha;

    // Step 2: download captcha image — same cookie jar via this.http.request()
    const imgUrl = `https://dropmail.me/captcha/image.png?_r=0&v=${encodeURIComponent(v)}&nonce=${encodeURIComponent(nonce)}&key=${encodeURIComponent(key)}&_sig=${encodeURIComponent(sig)}`;
    const imgRes = await this.http.request(imgUrl, { method: 'GET' });
    if (!imgRes.ok) return null;
    const imgBytes = await imgRes.arrayBuffer();

    // Step 3: OCR via PaddleOCR — plain fetch, no shared cookies
    let ocrText = null;
    const solvers = this._solvers || [paddleOcrSolver];
    for (const solver of solvers) {
      try {
        const result = await solver(imgBytes);
        if (result && result.trim()) {
          ocrText = result.trim();
          break;
        }
      } catch { /* ignore */ }
    }
    if (!ocrText) return null;

    // Step 4: submit solution — same cookie jar, form-encoded
    const formBody = new URLSearchParams({ response: ocrText, v, nonce, key, _sig: sig });
    const solRes = await this.http.request('https://dropmail.me/captcha/solution', {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: formBody.toString(),
    });
    let solData;
    try { solData = await solRes.json(); } catch { return null; }
    if (solData.result !== 'correct') return null;

    // Step 5: retry token generation with 90d — same cookie jar
    const tokenRes = await this.http.request(TOKEN_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ type: 'af', lifetime: '90d' }),
    });
    if (!tokenRes.ok) return null;
    const tokenData = await tokenRes.json();
    return tokenData.token || null;
  }

  /**
   * Generate token and create email address.
   * @returns {Promise<string>}
   */
  async generateEmail() {
    // Step 1: try 1d token (fast, no captcha)
    const tokenRes = await this.http.request(TOKEN_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ type: 'af', lifetime: '1d' }),
    });

    if (tokenRes.ok) {
      const tokenData = await tokenRes.json();
      this._token = tokenData.token;
    } else if (tokenRes.status === 402) {
      // Captcha required — attempt 90d flow
      let captchaData = {};
      try {
        const body = await tokenRes.json();
        captchaData = body.captcha || {};
      } catch { /* ignore */ }

      try {
        this._token = await this._solveCaptchaAndGetToken(captchaData);
      } catch {
        this._token = null;
      }

      if (!this._token) {
        // Fallback: retry 1d on possibly-refreshed session
        console.warn('Dropmail: captcha solve failed, retrying with 1d token');
        const fallbackRes = await this._fetch(TOKEN_URL, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ type: 'af', lifetime: '1d' }),
        });
        if (!fallbackRes.ok) throw new TempMailError(`Token generation failed: HTTP ${fallbackRes.status}`, this.name);
        const fallbackData = await fallbackRes.json();
        this._token = fallbackData.token;
      }
    } else {
      throw new TempMailError(`Token generation failed: HTTP ${tokenRes.status}`, this.name);
    }

    if (!this._token) throw new TempMailError('No token in response', this.name);
    this._endpoint = `https://dropmail.me/api/graphql/${this._token}`;

    // Create session and address
    const mutation = `
      mutation {
        introduceSession {
          id
          addresses {
            id
            address
            restoreKey
          }
        }
      }
    `;

    const data = await this._graphql(mutation);
    this._sessionId = data.introduceSession.id;

    const addresses = data.introduceSession.addresses;
    if (!addresses || addresses.length === 0) {
      throw new TempMailError('No address returned', this.name);
    }

    this._addressId = addresses[0].id;
    this._email = addresses[0].address;

    return this._email;
  }

  /**
   * @param {string} _email
   * @returns {Promise<Message[]>}
   */
  async getInbox(_email) {
    const query = `
      query($id: ID!) {
        session(id: $id) {
          mails {
            id
            fromAddr
            headerSubject
            receivedAt
            text
            html
            attachments {
              id
              name
              mime
              rawSize
            }
          }
        }
      }
    `;

    const data = await this._graphql(query, { id: this._sessionId });
    const mails = data.session?.mails || [];

    return mails.map(m => new Message({
      id: m.id,
      sender: m.fromAddr || 'unknown',
      subject: m.headerSubject || '',
      date: m.receivedAt,
    }));
  }

  /**
   * @param {string} messageId
   * @returns {Promise<MessageDetail>}
   */
  async readMessage(messageId) {
    // messageId format: "id:email" or just "id"
    const id = messageId.split(':')[0];

    const query = `
      query($id: ID!) {
        session(id: $id) {
          mails {
            id
            fromAddr
            headerSubject
            receivedAt
            text
            html
            attachments {
              id
              name
              mime
              rawSize
            }
          }
        }
      }
    `;

    const data = await this._graphql(query, { id: this._sessionId });
    const mails = data.session?.mails || [];
    const mail = mails.find(m => m.id === id);

    if (!mail) throw new NotFoundError(`Message ${id} not found`, this.name);

    return new MessageDetail({
      id: messageId,
      sender: mail.fromAddr || 'unknown',
      subject: mail.headerSubject || '',
      date: mail.receivedAt,
      bodyText: mail.text || '',
      bodyHtml: mail.html || '',
      attachments: (mail.attachments || []).map(a => ({
        filename: a.name,
        contentType: a.mime,
        url: '', // Dropmail doesn't provide direct URL, would need download API
      })),
    });
  }

  /**
   * Delete the email address.
   * @param {string} _email
   * @returns {Promise<boolean>}
   */
  async deleteEmail(_email) {
    if (!this._addressId) return false;

    const mutation = `
      mutation($input: DeleteAddressInput!) {
        deleteAddress(input: $input)
      }
    `;

    try {
      await this._graphql(mutation, {
        input: { addressId: this._addressId },
      });
      this._token = null;
      this._endpoint = null;
      this._sessionId = null;
      this._addressId = null;
      this._email = null;
      return true;
    } catch {
      return false;
    }
  }
}

File: rust/src/providers/dropmail.rs
use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::Mutex;

use crate::error::TempMailError;
use crate::httpclient::HttpClient;
use crate::models::{Message, MessageDetail};
use crate::providers::TempMailProvider;

const PADDLE_OCR_URL: &str = "https://mamamacjdjj-padle-predict.hf.space/predict";

/// PaddleOCR solver — 3 retries via HuggingFace space.
pub async fn paddle_ocr_solver(img_bytes: &[u8]) -> Option<String> {
    let ocr_client = reqwest::Client::new();
    let boundary = format!("DropmailOCR{}", chrono::Utc::now().timestamp_millis());
    for _ in 0..3 {
        let body = build_multipart(&boundary, img_bytes);
        let content_type = format!("multipart/form-data; boundary={}", boundary);
        let ocr_resp = match ocr_client
            .post(PADDLE_OCR_URL)
            .header("Content-Type", content_type.clone())
            .body(body)
            .send()
            .await
        {
            Ok(r) => r,
            Err(_) => continue,
        };
        let ocr_data: serde_json::Value = match ocr_resp.json().await {
            Ok(v) => v,
            Err(_) => continue,
        };
        if let Some(results) = ocr_data["results"].as_array() {
            if let Some(best) = results.first() {
                let confidence = best["confidence"].as_f64().unwrap_or(0.0);
                if confidence >= 0.7 {
                    if let Some(t) = best["text"].as_str() {
                        let trimmed = t.trim().to_string();
                        if !trimmed.is_empty() {
                            return Some(trimmed);
                        }
                    }
                }
            }
        }
    }
    None
}

/// Configuration for Dropmail provider.
#[derive(Debug, Clone, Default)]
pub struct DropmailConfig {}

#[derive(Debug)]
struct DropmailState {
    token: String,
    session_id: String,
    address_id: String,
    address: String,
}

pub struct Dropmail {
    client: HttpClient,
    state: Arc<Mutex<Option<DropmailState>>>,
    captcha_solvers: Vec<Arc<dyn Fn(&[u8]) -> Option<String> + Send + Sync>>,
}

impl std::fmt::Debug for Dropmail {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Dropmail")
            .field("state", &self.state)
            .field("captcha_solvers", &self.captcha_solvers.len())
            .finish()
    }
}

impl Dropmail {
    pub fn new(client: HttpClient, _config: Option<DropmailConfig>) -> Self {
        Self {
            client,
            state: Arc::new(Mutex::new(None)),
            captcha_solvers: Vec::new(),
        }
    }

    pub fn with_captcha_solvers(
        mut self,
        solvers: Vec<Arc<dyn Fn(&[u8]) -> Option<String> + Send + Sync>>,
    ) -> Self {
        self.captcha_solvers = solvers;
        self
    }
}

#[derive(Serialize)]
struct GenerateTokenReq<'a> {
    r#type: &'a str,
    lifetime: &'a str,
}

#[derive(Deserialize)]
struct GenerateTokenRes {
    token: Option<String>,
}

#[derive(Deserialize)]
struct CaptchaField {
    v: Option<String>,
    nonce: Option<String>,
    key: Option<String>,
    #[serde(rename = "_sig")]
    sig: Option<String>,
}

#[derive(Deserialize)]
struct TokenOrCaptchaRes {
    token: Option<String>,
    captcha: Option<CaptchaField>,
}

#[derive(Serialize)]
struct GraphqlReq<'a> {
    query: &'a str,
}

#[derive(Deserialize)]
struct GraphqlRes<T> {
    data: Option<T>,
    errors: Option<Vec<GraphqlError>>,
}

#[derive(Deserialize)]
struct GraphqlError {
    message: String,
}

#[derive(Deserialize)]
struct IntroduceSessionData {
    #[serde(rename = "introduceSession")]
    introduce_session: IntroduceSession,
}

#[derive(Deserialize)]
struct IntroduceSession {
    id: String,
    addresses: Vec<DropmailAddress>,
}

#[derive(Deserialize)]
struct DropmailAddress {
    id: String,
    address: String,
}

#[derive(Deserialize)]
struct SessionData {
    session: Option<Session>,
}

#[derive(Deserialize)]
struct Session {
    mails: Option<Vec<DropmailMail>>,
}

#[derive(Deserialize, Clone)]
struct DropmailMail {
    id: String,
    #[serde(rename = "fromAddr")]
    from_addr: Option<String>,
    #[serde(rename = "headerSubject")]
    header_subject: Option<String>,
    #[serde(rename = "receivedAt")]
    received_at: Option<String>,
    text: Option<String>,
    html: Option<String>,
    attachments: Option<Vec<serde_json::Value>>,
}

fn parse_date(s: &str) -> chrono::DateTime<chrono::Utc> {
    chrono::DateTime::parse_from_rfc3339(s)
        .map(|dt| dt.with_timezone(&chrono::Utc))
        .unwrap_or_else(|_| chrono::Utc::now())
}

/// Build a minimal multipart/form-data body for a single file field (no reqwest multipart feature needed).
fn build_multipart(boundary: &str, img_bytes: &[u8]) -> Vec<u8> {
    let mut body = Vec::new();
    let header = format!(
        "--{boundary}\r\nContent-Disposition: form-data; name=\"file\"; filename=\"cap.png\"\r\nContent-Type: image/png\r\n\r\n"
    );
    body.extend_from_slice(header.as_bytes());
    body.extend_from_slice(img_bytes);
    let footer = format!("\r\n--{boundary}--\r\n");
    body.extend_from_slice(footer.as_bytes());
    body
}

/// Percent-encode a string for use as a query parameter value.
fn pct_encode(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for b in s.bytes() {
        if b.is_ascii_alphanumeric() || matches!(b, b'-' | b'.' | b'_' | b'~') {
            out.push(b as char);
        } else {
            out.push_str(&format!("%{:02X}", b));
        }
    }
    out
}

/// Attempt captcha→90d flow. Returns token on success, None on any failure.
/// Uses a cookie-enabled reqwest::Client for Dropmail (same session).
async fn solve_captcha_and_get_token(
    captcha: CaptchaField,
    dropmail_client: &reqwest::Client,
    solvers: &[Arc<dyn Fn(&[u8]) -> Option<String> + Send + Sync>],
) -> Option<String> {
    let v     = captcha.v.unwrap_or_else(|| "3".into());
    let nonce = captcha.nonce.unwrap_or_default();
    let key   = captcha.key.unwrap_or_default();
    let sig   = captcha.sig.unwrap_or_default();

    // Step 2: download captcha image — same cookie session
    let img_url = format!(
        "https://dropmail.me/captcha/image.png?_r=0&v={}&nonce={}&key={}&_sig={}",
        pct_encode(&v), pct_encode(&nonce), pct_encode(&key), pct_encode(&sig),
    );
    let img_resp = dropmail_client.get(&img_url).send().await.ok()?;
    if !img_resp.status().is_success() { return None; }
    let img_bytes = img_resp.bytes().await.ok()?;

    // Step 3: solve via solver chain
    // Try user-provided sync solvers first
    let ocr_text = solvers.iter().find_map(|solver| solver(&img_bytes));
    let ocr_text = match ocr_text {
        Some(t) => t,
        None => {
            // Fall back to built-in async PaddleOCR
            paddle_ocr_solver(&img_bytes).await?
        }
    };

    // Step 4: submit solution — same cookie session, form-encoded
    let mut form_params = HashMap::new();
    form_params.insert("response", ocr_text.as_str());
    form_params.insert("v",        v.as_str());
    form_params.insert("nonce",    nonce.as_str());
    form_params.insert("key",      key.as_str());
    form_params.insert("_sig",     sig.as_str());

    let sol_resp = dropmail_client
        .post("https://dropmail.me/captcha/solution")
        .form(&form_params)
        .send()
        .await
        .ok()?;
    let sol_data: serde_json::Value = sol_resp.json().await.ok()?;
    if sol_data["result"].as_str() != Some("correct") {
        return None;
    }

    // Step 5: retry token generation with 90d — same cookie session
    let token_body = GenerateTokenReq { r#type: "af", lifetime: "90d" };
    let token_resp = dropmail_client
        .post("https://dropmail.me/api/token/generate")
        .json(&token_body)
        .send()
        .await
        .ok()?;
    if !token_resp.status().is_success() { return None; }
    let token_data: GenerateTokenRes = token_resp.json().await.ok()?;
    token_data.token
}

#[async_trait]
impl TempMailProvider for Dropmail {
    async fn generate_email(&self) -> Result<String, TempMailError> {
        // Cookie-enabled client for the entire Dropmail session
        let dropmail_client = reqwest::Client::builder()
            .cookie_store(true)
            .build()
            .map_err(|e| TempMailError::Api(format!("client build failed: {}", e)))?;

        // Step 1: try 1d token
        let resp_1d = dropmail_client
            .post("https://dropmail.me/api/token/generate")
            .json(&GenerateTokenReq { r#type: "af", lifetime: "1d" })
            .send()
            .await
            .map_err(|e| TempMailError::Api(format!("token request failed: {}", e)))?;

        let status = resp_1d.status();
        let token = if status.is_success() {
            let data: GenerateTokenRes = resp_1d
                .json()
                .await
                .map_err(|e| TempMailError::Api(format!("token parse failed: {}", e)))?;
            data.token.ok_or_else(|| TempMailError::Api("no token in response".into()))?
        } else if status.as_u16() == 402 {
            // Captcha required — attempt 90d flow
            let body: TokenOrCaptchaRes = resp_1d.json().await
                .unwrap_or(TokenOrCaptchaRes { token: None, captcha: None });

            let solved = if let Some(captcha) = body.captcha {
                solve_captcha_and_get_token(captcha, &dropmail_client, &self.captcha_solvers).await
            } else {
                None
            };

            if let Some(t) = solved {
                t
            } else {
                // Fallback: get 1d token via the main HttpClient
                eprintln!("Dropmail: captcha solve failed, falling back to 1d token");
                let fallback: GenerateTokenRes = self
                    .client
                    .post_json(
                        "https://dropmail.me/api/token/generate",
                        &GenerateTokenReq { r#type: "af", lifetime: "1d" },
                    )
                    .await
                    .map_err(|e| TempMailError::Api(format!("1d fallback failed: {}", e)))?;
                fallback.token.ok_or_else(|| TempMailError::Api("no token in fallback".into()))?
            }
        } else {
            return Err(TempMailError::Api(format!(
                "token generation failed: HTTP {}",
                status.as_u16()
            )));
        };

        let gql_url = format!("https://dropmail.me/api/graphql/{}", token);
        let mutation = r#"mutation { introduceSession { id addresses { id address } } }"#;
        let gql_req = GraphqlReq { query: mutation };
        let gql_res: GraphqlRes<IntroduceSessionData> = self
            .client
            .post_json(&gql_url, &gql_req)
            .await
            .map_err(|e| TempMailError::Api(format!("GraphQL introduceSession failed: {}", e)))?;

        if let Some(errors) = gql_res.errors {
            if !errors.is_empty() {
                return Err(TempMailError::Api(format!("GraphQL error: {}", errors[0].message)));
            }
        }

        let session_data = gql_res
            .data
            .ok_or_else(|| TempMailError::Api("no data in GraphQL response".into()))?;
        let session = session_data.introduce_session;
        let address = session
            .addresses
            .into_iter()
            .next()
            .ok_or_else(|| TempMailError::Api("no addresses returned".into()))?;

        let mut state = self.state.lock().await;
        *state = Some(DropmailState {
            token,
            session_id: session.id,
            address_id: address.id,
            address: address.address.clone(),
        });

        Ok(address.address)
    }

    async fn get_inbox(&self, _email: &str) -> Result<Vec<Message>, TempMailError> {
        let state = self.state.lock().await;
        let state = state.as_ref().ok_or_else(|| {
            TempMailError::Api("Must call generate_email first".into())
        })?;

        let query = format!(
            r#"query {{ session(id: "{}") {{ mails {{ id fromAddr headerSubject receivedAt }} }} }}"#,
            state.session_id
        );
        let gql_req = GraphqlReq { query: &query };
        let gql_res: GraphqlRes<SessionData> = self
            .client
            .post_json(
                &format!("https://dropmail.me/api/graphql/{}", state.token),
                &gql_req,
            )
            .await
            .map_err(|e| TempMailError::Api(format!("GraphQL getInbox failed: {}", e)))?;

        let mails = gql_res
            .data
            .and_then(|d| d.session)
            .and_then(|s| s.mails)
            .unwrap_or_default();

        Ok(mails
            .into_iter()
            .map(|m| Message {
                id: m.id,
                sender: m.from_addr.unwrap_or_default(),
                subject: m.header_subject.unwrap_or_default(),
                date: m.received_at.as_deref().map(parse_date).unwrap_or_else(chrono::Utc::now),
            })
            .collect())
    }

    async fn read_message(&self, message_id: &str) -> Result<MessageDetail, TempMailError> {
        let state = self.state.lock().await;
        let state = state.as_ref().ok_or_else(|| {
            TempMailError::Api("Must call generate_email first".into())
        })?;

        let query = format!(
            r#"query {{ session(id: "{}") {{ mails {{ id fromAddr headerSubject receivedAt text html attachments }} }} }}"#,
            state.session_id
        );
        let gql_req = GraphqlReq { query: &query };
        let gql_res: GraphqlRes<SessionData> = self
            .client
            .post_json(
                &format!("https://dropmail.me/api/graphql/{}", state.token),
                &gql_req,
            )
            .await
            .map_err(|e| TempMailError::Api(format!("GraphQL readMessage failed: {}", e)))?;

        let mails = gql_res
            .data
            .and_then(|d| d.session)
            .and_then(|s| s.mails)
            .unwrap_or_default();

        mails
            .into_iter()
            .find(|m| m.id == message_id)
            .map(|mail| MessageDetail {
                message: Message {
                    id: mail.id,
                    sender: mail.from_addr.unwrap_or_default(),
                    subject: mail.header_subject.unwrap_or_default(),
                    date: mail.received_at.as_deref().map(parse_date).unwrap_or_else(chrono::Utc::now),
                },
                body_text: mail.text,
                body_html: mail.html,
                attachments: mail.attachments.unwrap_or_default(),
            })
            .ok_or(TempMailError::NotFound)
    }

    async fn delete_email(&self, _email: &str) -> Result<bool, TempMailError> {
        let state = self.state.lock().await;
        let state = state.as_ref().ok_or_else(|| {
            TempMailError::Api("Must call generate_email first".into())
        })?;

        let query = format!(
            r#"mutation {{ deleteAddress(input: {{ addressId: "{}" }}) }}"#,
            state.address_id
        );
        let gql_req = GraphqlReq { query: &query };
        let _gql_res: serde_json::Value = self.client
            .post_json(
                &format!("https://dropmail.me/api/graphql/{}", state.token),
                &gql_req,
            )
            .await
            .map_err(|e| TempMailError::Api(format!("GraphQL deleteAddress failed: {}", e)))?;

        Ok(true)
    }
}
